src/Orleans/Utils/Factory.cs

```diff
@@ -0,0 +1,36 @@
+﻿namespace Orleans
+{
+    /// <summary>
+    /// Creates an instance of <typeparamref name="TInstance"/>.
+    /// </summary>
+    /// <typeparam name="TInstance"></typeparam>
+    /// <returns></returns>
+    public delegate TInstance Factory<out TInstance>();
```

Comment:
Generally speaking for parameterized factories it makes sense, but the this Factory registration is nothing different from a Transient registration, since DI will always return a new instance and accepts no parameters.

What I am missing from this factory delegate pattern is that I can't see why we don't include the ```ServiceProvider``` as a default first parameter in all the delegates to make it more friendly with DI containers. This way none of the factory functions have a way to use ```ActivatorUtilities.CreateInstance``` to pass in a ```IServiceProvider``` instance.

Since DI Abstractions already having the ```ObjectFactory``` delegate via the ```ActivatorUtilities.CreateFactory``` method, it can be a nice addition to it, hence we should call the same ```ObjectFactory``` instead of ```Factory``` ?